File: DAYS/Day33-Alien Dictionary/alienOrder.js
//Graphs- Alien Dictionary Order
//Given a list of words from the alien language dictionary, where the words are sorted lexicographically by the rules of this new language, derive the order of letters in this language.

//Example 1: Input: words = ["wrt","wrf","er","ett","rftt"] Output: "wertf"
//Example 2: Input: words = ["z","x"] Output: "zx"
//Example 3: Input: words = ["z","x","z"] Output: "" Explanation: The order is invalid, so return "".
// Different Approach: 
// 1. DFS Topological Sort
// 2. BFS Topological Sort (Kahn's Algorithm)

// Solution: BFS Topological Sort (Kahn's Algorithm)
// Leetcode: https://leetcode.com/problems/alien-dictionary/solutions/1678823/javascript-bfs-topological-sort-kahn-s-algorithm-with-detailed-explanation/?orderBy=most_relevant

function alienOrder(words){
    if(!words ||words.length ===0) return "";

    // Step 1: Create the graph
    const graph = new Map();
    const indegree = new Map();
    for(let word of words){
        for(let ch of word){
            if(!graph.has(ch)) graph.set(ch,new Set());
            if(!indegree.has(ch)) indegree.set(ch,0);
        }
    }
    // Step 2: Build the graph
    for(let i=0;i<words.length-1;i++){
        const w1=words[i];
        const w2=words[i+1];
        if(w1.length>w2.length && w1.startsWith(w2)) return ""; // Invalid case
        const minLength=Math.min(w1.length,w2.length);
        for(let j=0;j<minLength;j++){
            if(w1[j]!==w2[j]){ // Find the first different character
                if(!graph.get(w1[j]).has(w2[j])){   // Avoid duplicate edges
                    graph.get(w1[j]).add(w2[j]); // Add edge
                    indegree.set(w2[j],indegree.get(w2[j])+1); // Increase indegree
                    break; // Only the first different character between the two words will help us find the order
                }
            }
        }
    }
    // Step 3: Topological Sort using BFS (Kahn's Algorithm)
    const queue = [];
    for(let [key,value] of indegree.entries()){
        if(value===0) queue.push(key); // Start with nodes with 0 indegree
    }   
    const result = [];
    while(queue.length>0){
        const ch = queue.shift();
        result.push(ch);
        for(let neighbor of graph.get(ch)){
            indegree.set(neighbor,indegree.get(neighbor)-1); // Decrease indegree
            if(indegree.get(neighbor)===0) queue.push(neighbor); // If indegree becomes 0, add to queue
        }
    }
    return result.length===indegree.size ? result.join('') : ''; // If result length is not equal to indegree size, there is a cycle
}

// Test Cases
console.log(alienOrder(["wrt","wrf","er","ett","rftt"])); // Output: "wertf"
console.log(alienOrder(["z","x"])); // Output: "zx"
console.log(alienOrder(["z","x","z"])); // Output: ""
// Best Approach: BFS Topological Sort (Kahn's Algorithm)
// Complexity Analysis: Time Complexity- O(v+e) , Space Complexity- O(v+e) 
// where v is the number of unique characters and e is the number of edges in the graph.

File: DAYS/Day33-Alien Dictionary/alienOrder.py
# Graphs - Alien Dictionary
# Given a list of words from the alien language's dictionary, where the words are sorted lexicographically 
# by the rules of this new language, derive the order of letters in this language.
# Example 1: Input: ["wrt","wrf","er","ett","rftt"] Output: "wertf"
# Example 2: Input: ["z","x"] Output: "zx"
# Example 3: Input: ["z","x","z"] Output: "" (invalid order)
# Different Aproaches:
# 1. Topological Sort using Kahn's Algorithm (BFS)
# 2. Topological Sort using DFS

# Topological Sort solution using Kahn's Algorithm (BFS)
# Leetcode: https://leetcode.com/problems/alien-dictionary/

from collections import deque
class Solution:
    def alienOrder(words):
        if not words:
            return ""
        graph={c :set() for word in words for c in word} # adjacency list
        indegree={c:0 for c in graph} # indegree count

        for i in range(len(words)-1): 
            w1,w2=words[i],words[i+1]  #w1 is before w2
            if len(w1)>len(w2) and w1.startswith(w2): 
                return "" # invalid case
            for j in range(min(len(w1),len(w2))):
                if w1[j]!=w2[j]: # first different char
                    if w2[j] not in graph[w1[j]]:
                        graph[w1[j]].add(w2[j]) # add edge
                        indegree[w2[j]]+=1 # increase indegree
                    break
        queue=deque([c for c in indegree if indegree[c]==0]) # all chars with 0 indegree
        order=[]
        while queue: 
            c=queue.popleft() # char with 0 indegree
            order.append(c) # add to order
            for nei in graph[c]: # reduce indegree of neighbors
                indegree[nei]-=1
                if indegree[nei]==0:
                    queue.append(nei) # add to queue if indegree is 0
        if len(order)==len(indegree): # cycle detected
            return "".join(order) # valid order
        else:
            return "" # invalid order due to cycle

# Test Cases
print(Solution.alienOrder(["wrt","wrf","er","ett","rftt"])) # Output: "wertf"
print(Solution.alienOrder(["z","x"])) # Output: "zx"
    
# Best Approach: Topological Sort using Kahn's Algorithm (BFS)
# Complexity Analysis: Time: O(V + E), Space: O(V + E)  
# where V is the number of unique characters and E is the number of edges in the graph.